Coingecko feature branch

JakubPluta opened the pull request:
I created separate coingecko module inside cryptocurrency. Also I added seperate ta controller for crypto. I propose to use ta controller separately in each crypto module which provide prices data. E.g coingecko have methods that fetch prices, so it will have ta available, in future coinpaprika also will have ta controller and so on..

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+    """
+    parser = argparse.ArgumentParser(
+        add_help=False,
+        prog="base_info",
```
name of program should be the command used by the user, which in this case is just "info"

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+    """
+    parser = argparse.ArgumentParser(
+        add_help=False,
+        prog="web",
```
You forgot to update this one to social

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+    """
+    parser = argparse.ArgumentParser(
+        add_help=False,
+        prog="market_data",
```
Should be just "market" here I think

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+""" pycoingecko_api """
```
Apart from TA all of these worked well for me. Amazing work.

The only thing I would have done is add a better description to the commands, so that when the user does "market -h" there is more information on what to expect of type of data, e.g. the rows elements.

Since this is a dataframe, you can use this:
```
", ".join(df.index)
```

on your notebook, and then just copy-paste in there. At least that's how I did it for the Financial Analysis stuff

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_overview_controller.py:
```diff
@@ -0,0 +1,267 @@
+        "help",
+        "q",
+        "quit",
+        "global",
```
Some of these commands are missing a `print("")` at the end.

These are: global, coins, defi, top_defi, top_dex, top_nft, nft_market, exchanges, ex_rates, platforms, products, indexes, derivatives,

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_overview_controller.py:
```diff
@@ -0,0 +1,267 @@
+        print("   most_voted      show most voted coins on CoinGecko")
+        print("   most_visited    show most visited coins on CoinGecko")
+        print("   top_volume      show coins with highest volume on CoinGecko")
+        print("   recently        show recently added on CoinGecko")
```
I think in this command you should get rid of the URL, or post it in a different way. 

E.g. could be another command for the user to access this url, like:
```
url Fabric
```
And you get: https://www.coingecko.com/en/coins/fabric

This is so that we avoid the columns getting messed up due to size of table

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+
+    """
+    parser = argparse.ArgumentParser(
+        prog="stablecoins",
```
yfarms*

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+        if not ns_parser:
+            return
+
+        df = gecko.get_yield_farms()
```
In this one the table is just too long. In my case I couldn't even scroll through the header.

An idea here would be to allow the user to specify the order of the data they want to see and the number of elements.

As default you could do something like showing first 10 and ordering by one of those last 2 columns.

Also, if all the columns contain the same word, e.g. Hourly/Yearly, it is probably worth incorporating it on the column name, so that we avoid likelihood of overrunning screen.

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+        if not ns_parser:
+            return
+
+        df = gecko.get_top_defi_coins()
```
Add flag with number to show, by default I would just add top 10/20. The idea is that the user can get data fast and move fast. So, whenever possible we want the user to be able to reach the data he wants to see with the minimum overhead.

In this case I think is very unlikely that someone care about the coins that are in top 19 or 56. However, if it is in top 10 is probably meaningful.

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+        if not ns_parser:
+            return
+
+        df = gecko.get_companies_assets(endpoint=ns_parser.coin)
```
Did you do something different here?

Asking because the table shows like this:
<img width="1177" alt="Captura de ecrã 2021-06-28, às 02 25 53" src="https://user-images.githubusercontent.com/25267873/123566729-4e260180-d7b8-11eb-8b98-6180dbd35c63.png">

Where in 1 hand the table doesn't overrun the screen, but in the other hand the link is not cliable anymore as there are new lines in betwen.

JakubPluta replied:
Yes, in some data frames which has big width I added a method, that wraps long text into new line, base on given len. So this way I could fit width of displayed table into terminal width. But question, what to you think it's better idea for that ? 
1. Remove all the urls from tables (issue with too-long columns will be fixed)
2. Like in you example, just split to long text into new lines (inside column)
What do you think ?

DidierRLopes replied:
So, my opinion on that is that:
1. We want the table to be displayed nicely to the user
2. We want the link to be clickable

In this case you solve 1 at the expense of 2. Which works, but the issue is that the link is no longer clicable, and if you want to access it you need to copy-paste for as many lines as the link contains, which is not ideal.

Hence, due to that I would say to not display columns by default, and have a flag to "-l" or "--links"  and if the user selects such bool flag, we would display a new dataframe with columns Rank, Company, Url.

What do you think?

This is just myself putting me on the user side of things, and thinking about what I would find useful

JakubPluta replied:
Yea, agree. From user perspective it should be clicable and pretty. So the idea for views that display urls, which are very long should have this kind of flag (probably there will be few more views which can have simillar behaviour). But let's try with --links flag
Thank you for advice

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+
+    """
+    parser = argparse.ArgumentParser(
+        prog="top_categories",
```
categories*

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_view.py:
```diff
@@ -0,0 +1,1570 @@
+        if not ns_parser:
+            return
+
+        df = gecko.get_top_crypto_categories()
```
One idea for these tables is to try to shorten the zeros of the billions. You can see how it's done in the fundamental analysis menu.

Instead of displaying `1,123,123,123 $`. We do `1.123 B$`

DidierRLopes commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_overview_controller.py:
```diff
@@ -0,0 +1,267 @@
+        print("   quit            quit to abandon program")
+        print("")
+        print("Coingecko Overview: ")
+        print("   global          show global crypto market info")
```
Since all the ones below contain a "show" I would strip the show word from all of them as they don't add anything

DidierRLopes reviewed (commented):
Overall, what an amazing PR. This is massive, amazing work. I tested all functionalities and they all work for me, except Technical Analysis (for the reasons I mentioned in here).

In general, the suggestions I gave can be sum up to:
* Refactor coingecko to 1 single controller
* Add more descriptions to the commands, so that the user knows what we're talking about before running it (this would be added also in README)
* Try to shorten the tables
   * For both columns, by either removing hyperlinks, or modifying big numbers to their shorter versions
   * For rows, by allowing the user to specify how many rows he wants to see along with what the will be sorted by
* Add new line between commands used
* Incorporate loaded coin to the help menu, to remove "noise" from user's screen
*  Confirm the usage name matches the function the user sees.

Extremely impressed by this, particularly by the documentation. Extremely good. Glad to have you onboard!

DidierRLopes commented on gamestonk_terminal/cryptocurrency/ta_crypto_controller.py:
```diff
@@ -0,0 +1,98 @@
+)
+
+
+class CryptoTechnicalAnalysisController(TAC):
```
Why did we add a TA specifically for crypto? Since it uses the same functionalities, I'm sure we're able to somehow manage it from the TA perspective of things. That would be ideal, as it would remove duplicate code.

DidierRLopes commented:
> I created separate coingecko module inside cryptocurrency. Also I added seperate ta controller for crypto. I propose to use ta controller separately in each crypto module which provide prices data. E.g coingecko have methods that fetch prices, so it will have ta available, in future coinpaprika also will have ta controller and so on..

We can discuss about the Ta_controller on a call. But the way I see it is that the TA is agnostic to what the data is, it just needs to process datapoints in his own particular format. So the coingecko, or coinpaprika will just need to convert their dataframes into the type accepted by this TA. Unless I'm missing something!

JakubPluta commented on docs/make.bat:
```diff
@@ -1,35 +1,35 @@
-
-:end
-popd
+@ECHO OFF
```
To be honest, I have no idea. My bad, I even didn't notice that I committed it. My IDEE  plays tricks on me.

DidierRLopes replied:
No problem, I was just wondering what it was because I didn't know :p

DidierRLopes commented:
The tests on your PR are failing not due to your code. Not sure what triggered this, probably a pylint update.

Anyway, I'm gonna push a fix for these issues, I was also getting them on my PR. Once I do you can merge main to your feature branch and then we can re-run the tests.

DidierRLopes commented:
> The tests on your PR are failing not due to your code. Not sure what triggered this, probably a pylint update.
> 
> Anyway, I'm gonna push a fix for these issues, I was also getting them on my PR. Once I do you can merge main to your feature branch and then we can re-run the tests.

I just pushed these changes. So feel free to merge main onto your branch whenever you're ready!

JakubPluta commented:
> The tests on your PR are failing not due to your code. Not sure what triggered this, probably a pylint update.
> 
> Anyway, I'm gonna push a fix for these issues, I was also getting them on my PR. Once I do you can merge main to your feature branch and then we can re-run the tests.

Sure, thanks for that! So very slowly I am doing some refactoring and apply changes according to your review. BTW, I will need some advice regarding this ta controller. I tried to add context to controller to have like
`(crypto)(cg)(ta)` and when doin `q` it should return to (crypto)(cg), but it doesn't work as expected. Is it doable with current architecture in cases when we have 3-level menu ?

JakubPluta commented on gamestonk_terminal/cryptocurrency/coingecko/pycoingecko_coin_controller.py:
```diff
@@ -0,0 +1,254 @@
+            print("")
+
+    # pylint: disable=inconsistent-return-statements
+    def call_ta(self, other_args):
```
@DidierRLopes  I am trying to refactore it, but in the end even if I have context defined, still `q` returns do `(crypto)` menu instead to `(crypto)(cg)`. Is it possible to have executed `ta_controller.menu` at the level of coingecko ? (So it means 3rd level of menu depth - `(crypto)(cg)(ta)`.

DidierRLopes commented:
> > The tests on your PR are failing not due to your code. Not sure what triggered this, probably a pylint update.
> > Anyway, I'm gonna push a fix for these issues, I was also getting them on my PR. Once I do you can merge main to your feature branch and then we can re-run the tests.
> 
> Sure, thanks for that! So very slowly I am doing some refactoring and apply changes according to your review. BTW, I will need some advice regarding this ta controller. I tried to add context to controller to have like
> `(crypto)(cg)(ta)` and when doin `q` it should return to (crypto)(cg), but it doesn't work as expected. Is it doable with current architecture in cases when we have 3-level menu ?

It should be, let me have a look at this for you!

DidierRLopes commented:
> > > The tests on your PR are failing not due to your code. Not sure what triggered this, probably a pylint update.
> > > Anyway, I'm gonna push a fix for these issues, I was also getting them on my PR. Once I do you can merge main to your feature branch and then we can re-run the tests.
> > 
> > 
> > Sure, thanks for that! So very slowly I am doing some refactoring and apply changes according to your review. BTW, I will need some advice regarding this ta controller. I tried to add context to controller to have like
> > `(crypto)(cg)(ta)` and when doin `q` it should return to (crypto)(cg), but it doesn't work as expected. Is it doable with current architecture in cases when we have 3-level menu ?
> 
> It should be, let me have a look at this for you!

I'll send you a patch with this over discord :)

JakubPluta commented:
Hi @DidierRLopes. I followed all your comments from review and refactored the code. I also added a lot of new parameters that allow the user to sort, select the top n number of records or display source links. I also updated the README.
If you will find time would be great to have a 2nd review :)

DidierRLopes commented:
Sorry, only had time to go through it now!

Amazing work! So excited to have this!!

In the main menu I would do:
```
Cryptocurrency:
   help            show this menu again
   q                 quit this menu, and shows back to main menu
   quit             quit to abandon program

>  cg               CoinGecko overview (market statistics) and coin menu
>  cmc            Coinmarketcap menu
>  bin              Binance menu with order book, candles, ta.. 
```

If going to crypto/bin/candle without loading it fails badly, probably add a check at the beginning of the command, to check that there's a crypto loaded.
```
 (crypto)>(bin)> candle
Traceback (most recent call last):
  File "terminal.py", line 540, in <module>
    terminal()
  File "terminal.py", line 513, in terminal
    process_input = t_controller.switch(an_input)
  File "terminal.py", line 207, in switch
    return getattr(
  File "terminal.py", line 290, in call_crypto
    return crypto_controller.menu()
  File "/Users/DidierRodriguesLopes/Documents/GamestonkTerminal/gamestonk_terminal/cryptocurrency/crypto_controller.py", line 116, in menu
    process_input = crypto_controller.switch(an_input)
  File "/Users/DidierRodriguesLopes/Documents/GamestonkTerminal/gamestonk_terminal/cryptocurrency/crypto_controller.py", line 64, in switch
    return getattr(
  File "/Users/DidierRodriguesLopes/Documents/GamestonkTerminal/gamestonk_terminal/cryptocurrency/crypto_controller.py", line 87, in call_bin
    if binance_controller.menu():
  File "/Users/DidierRodriguesLopes/Documents/GamestonkTerminal/gamestonk_terminal/cryptocurrency/binance/binance_controller.py", line 157, in menu
    process_input = binance_controller.switch(an_input)
  File "/Users/DidierRodriguesLopes/Documents/GamestonkTerminal/gamestonk_terminal/cryptocurrency/binance/binance_controller.py", line 59, in switch
    return getattr(
  File "/Users/DidierRodriguesLopes/Documents/GamestonkTerminal/gamestonk_terminal/cryptocurrency/binance/binance_controller.py", line 82, in call_candle
    binance_model.show_candles(
  File "/Users/DidierRodriguesLopes/Documents/GamestonkTerminal/gamestonk_terminal/cryptocurrency/binance/binance_model.py", line 199, in show_candles
    f"{coin+currency} from {candles_df.index[0].strftime('%Y/%m/%d')} to "
TypeError: unsupported operand type(s) for +: 'NoneType' and 'NoneType'
```

DidierRLopes commented:
Also, when doing "candles" in binance menu. There's 2 new lines instead of just 1 :)

JakubPluta commented:
@DidierRLopes Thank you very much for the review :) 

Once again I missed some small details. But already pushed changes. 

Check for candles and orders is added: 

![image](https://user-images.githubusercontent.com/275820/124374559-9358a080-dc9c-11eb-8355-f539d421e732.png)

And also I removed headers from main crypto menu `print_help` 

![image](https://user-images.githubusercontent.com/275820/124374592-d0249780-dc9c-11eb-9e62-4deeae237c09.png)

If all good, I think we can merge it 👍 

Once more, thanks for your time!

DidierRLopes reviewed (approved):
THIS IS THE WAY!! Amazing PR! Good work!
